Extract queue_job module from connector

guewen opened the pull request:
* Replace ConnectorSession by odoo.api.Environment
* Rework job classes
* Jobs are now applied on Model methods

```python
  class MyModel(models.Model):
     _name = 'my.model'

     @api.multi
     @job
     def my_method(self, a, k=None):
         _logger.info('executed with a: %s and k: %s', a, k)


  class MyOtherModel(models.Model):
      _name = 'my.other.model'

      @api.multi
      def button_do_stuff(self):
          self.env['my.model'].with_delay().my_method('a', k=2)
```

See history on https://github.com/OCA/connector/pull/212

sbidoul committed:
[FIX] connector-runner: do not attempt to auto-create/auto-configure channels

What we receive from the queue are channel names and that's all.
My original idea was to let applications auto-create temporary channels
which could be useful for sequential channels.
But this needs more thinking.

sbidoul committed:
[IMP] connector-runner: more new-style classes

sbidoul committed:
[IMP] connector-runner: do not NOTIFY when deleting done jobs

This NOTIFY is not necessary and makes the purge of the job queue very slow.
Thanks to @acsonelmi for the suggestion.

sbidoul committed:
[FIX] connector-runner: use string type for datetime

Odoo indeed adapts the psycopg2 type mappings to use strings for datetime.

sbidoul committed:
[IMP] connector-runner: tweak logging

Commit:
Add tests on jobs and channels models

Commit:
Allow to assign a default channel on @job functions

When the @job is registered and created in the database, it lookups its channel
and creates it if it don't exist yet.

sbidoul committed:
[IMP] connector-runner: job id is a perfectly good proxy for seq

sbidoul committed:
[FIX] connector-runner: fix time zone issue when enqueing jobs

sbidoul committed:
[IMP] connector-runner: remove ODOO_CONNECTOR_ENABLE environment variable

The presence of ODOO_CONNECTOR_CHANNELS is enough to make the runner
start in place of the workers.

sbidoul committed:
[IMP] connector-runner: handle changes in job scheduling properties

sbidoul committed:
[IMP] connector-runner: remove obsolete TODO

Since the runner is now a thread in the Odoo main process,
such a situation should occur only with a pathologically wrong
configuration, so no need to worry about that... unless proven
otherwise ;)

sbidoul committed:
[WIP] connector-runner: sphinx documentation

sbidoul committed:
[IMP] connector-runner: refactor handling of 'now'

So the channel module does not make assumptions on the type
of the eta field, so we use the odoo-ish way to get now in
a database-compatible format, so it is more easily testable.

sbidoul committed:
[IMP] connector-runner: doctest formatting

sbidoul committed:
[IMP] connector-runner: support missing root prefix in channel names

sbidoul committed:
[IMP] connector-runner: high level doctest for the ChannelManager

sbidoul committed:
[IMP] connector-runner: tiny refactoring in channel name config

sbidoul committed:
[FIX] connector-runner: documentation typos

Commit:
Release 3.1.0

Commit:
Wrong arguments position in a log message

sbidoul committed:
[IMP] connector-runner: graceful stop mechanism

The previous version was using a daemon thread, which as more or less
ok, except it could enqueue jobs while the server was stopping, and such
jobs would never start.

yvaucher committed:
Add bug tracker link on README.rst

bguillot committed:
[IMP] add attribute 'exported_fields'on Mapper classes to list the exported fields

bguillot committed:
[FIX] remove useless print

bguillot committed:
[ADD] support function on direct fields

bguillot committed:
[IMP] rename exported_fields to changed_by_fields

bguillot committed:
[ADD] comment

Commit:
See if we can remove install_in_connector() by checking in ir.module.module with caching of the result

Commit:
Update doc

JosDeGraeve committed:
[IMP] Add retry pattern for jobs

This adds a retry pattern for queue jobs.
see OCA/connector-magento#26

Commit:
Fix, the method 'module_uninstall' invalid the ormcache of is_module_installed

JosDeGraeve committed:
FIX: sort the retry dictionairy by retry count

Sort the retry dict by key ( the retry count ) instead of by value

bguillot committed:
[FIX] change list to set and fix nested direct functions

Commit:
improve comparaison test

bguillot committed:
[IMP] add tests for _changed_by_fields attribute

bguillot committed:
[FIX] use staticmethod to keep _mapping_field_name() inside the metaclass

bguillot committed:
[FIX] _mapping_field_name()

bguillot committed:
[FIX] useless blank line

Commit:
Merge pull request #70 from yvaucher/8.0-add-bug-tracker-link

Add bug tracker link on README.rst

Commit:
no need to call self.env['ir.module.module'] in ir.module.module model

Commit:
no check a boolean value with None value

bguillot committed:
[FIX] clean code

bguillot committed:
[FIX] rename method

bguillot committed:
[FIX] flake8

guewen committed:
Set Odoo path from travis environ

Commit:
Fix is_module_installed function, check if one model of the module connector exist before call the method define by module connector

Commit:
Fix unittest, when test is called by odoo, the module are not in the state 'installed' but 'to install'

bguillot committed:
[FIX] get the exact field name (mapping the closure entries)

Commit:
Invalid cache in write method only if the state is in the value to update

Commit:
Update docstrings by Guewen (c2c)

guewen committed:
Keep the retry pattern on the @job definition rather than in jobs

It avoids to copy and to store the pattern for every job when all jobs of one
function will probably share the same pattern.

guewen committed:
Use seconds of the higher retry count of the retry pattern

So that the last count is repeated indefinitely.
For the following pattern:
{1: 10*60,
 5: 20*60,
 10: 60*60}
The retries 1 to 4 will be postponed 10 minutes later.
The retries 5 to 9 will be postponed 20 minutes later.
The retries 10 to infinite will be postponed 1 hour later.

ccomb committed:
Implemented a default binder to avoid reimplementing it in each connector

ccomb committed:
Mention the default implementation is the doc

guewen committed:
Use odoo-sphinx-autodoc to build correctly the doc

Otherwise the auto api docs are not generated

guewen committed:
New test: check if the cache is different for different modules

guewen committed:
Rename module.py to ir_module_module.py

guewen committed:
Test the cache invalidation on 'base' instead of 'connector'

Because the check uses a fast-path for the connector so it does not reach the
cache.

guewen committed:
Remove useless trick in is_module_installed

This code is never reached when we check the installation state with the
high-level function is_module_installed for the module 'connector' because it
directly looks for a model in the registry to check the installation state

Commit:
Merge pull request #74 from jssuzanne/8.0

Deprecate install_in_connector()

pedrobaeza committed:
Merge pull request #77 from guewen/doc-build-no-buildout

Do not use buildout to build the doc in travis

bealdav committed:
[IMP] job view

pedrobaeza committed:
Merge pull request #78 from akretion/8-job-form-view

[IMP] job view

Commit:
Merge pull request #69 from acsone/8.0-runnerstop-sbi

[IMP] connector-runner: graceful stop mechanism

Commit:
Merge pull request #73 from akretion/8.0_smart-export

adds an attribute changed_by_fields on the Mapper classes

Allowing to avoid to call a Mapper if we know that we none of the field triggers a change.

Commit:
Remove field queue_job.retry_pattern

sbidoul committed:
[FIX] common.DB is gone in 8.0 stable

mathi123 committed:
[DOC] small improvent

Commit:
manage retryable job with runner

Commit:
Add an index on queue_job.worker_id

This column appears often in the where clauses with the old workers.

pedrobaeza committed:
Merge pull request #89 from guewen/8.0-connector-worker_id-index

Add an index on queue_job.worker_id

Commit:
Correct flake8 warning: long line

Commit:
Slight changes in the docstrings

Commit:
Binder.to_openerp() always returns a recordset

Commit:
Add line in the changes file

Commit:
Add a line in the changes file

LeartS committed:
Mapper: add modifier to follow m2o relations

Does not do particular checks to see if each "step" is a
valid m2o field, relies on the standard and familiar
Odoo recordset errors instead.

LeartS committed:
Add simple test case and fix PEP8

Commit:
Move fields to avoid jobs view overflows

The job function, channel and worker fields are too large for a 6 column
layout. Job function and channel make the fields overflow of the form and the
worker fields is wrapped over 3 lines

Commit:
A None capacity means infinite. Fix logging error

Fixes #95

Commit:
Do not pass self as 'message' of set_pending()

Fixes #96

Commit:
[FIX] seconds defaults to RETRY_INTERVAL when no retry_pattern is specified (to avoid timedelta error when seconds is None) in job.py.

nhomar committed:
Merge pull request #101 from guewen/8.0-runjobcontroller-result

Do not pass self as 'message' of set_pending()

Commit:
[FIX] retry_pattern in job decorator.

pedrobaeza committed:
Merge pull request #98 from guewen/none-capacity-error

A None capacity means infinite. Fix logging error

Commit:
Merge pull request #105 from taktik/8.0-fix-retry-pattern-decorator

[FIX] retry_pattern in job decorator.

Commit:
Merge pull request #103 from taktik/8.0-fix-timedelta

[FIX] seconds defaults to RETRY_INTERVAL when no retry_pattern

Commit:
create environment with a classmethod

Commit:
FIX: pep8

Commit:
Added unittests

Commit:
FIX: Make sure kwargs is instantiated

Commit:
FIX: documentation / ADD: authors and changes

Commit:
FIX: Typo in CHANGES.rst / DOC: py:meth instead py:class

flotho committed:
Add OCA icon

oca-transbot committed:
OCA Transbot updated translations from Transifex

gurneyalex committed:
Merge pull request #97 from guewen/8.0-job-view-overflow

Move fields to avoid jobs view overflows

Commit:
Release 3.2.0

Commit:
Correct the authors list: missing company

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
[IMP] improve job runner documentation

sbidoul committed:
[IMP] mention web_kanban in the --load option as it is the odoo default

Plus various cosmetic improvements.

sbidoul committed:
[IMP] answer another jobrunner FAQ in the documentation

Commit:
Add 'mock_job_delay_to_direct' to ease tests on jobs

This context manager allows to execute a job function synchronously when it
should normally have been delayed in a asynchronous job.

This is really useful for writing tests that check the flow of a
synchronisation without having to deal with the jobs execution mechanisms.

It is used in the OCA/connector-magento project and ought to be used by other
projects.

lepistone committed:
Merge pull request #119 from MindAndGo/8.0-product-template-view

[REF]8.0 product template view

Commit:
Allow to define seconds when raising RetryableJobError

The job will be retried after the given number of seconds if the exception is
raised with a defined 'seconds' argument.

Commit:
Allow to ignore the retry counter when raising RetryableJobError

Commit:
Exception should be raised with arguments

sbidoul committed:
[UPD] prefix versions with 8.0

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
Add WooCommerce connector link in documentation

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Make modules uninstallable

Commit:
Check if a module is installed from the registry

Instead of `ir.module.module`. Simplifies as we never hit the database nor have
to maintain a cache.

Also, it fixes an issue when we run tests directly at the installation of a
module, during the tests the previous implementation was considering the module
as uninstalled instead of installed.

Change proposed @bealdav. Thanks!

Fixes #85

lmignon committed:
Merge pull request #124 from guewen/retryable-exception-seconds

Allow to define seconds and ignore retries when raising RetryableJobError

oca-transbot committed:
OCA Transbot updated translations from Transifex

npiganeau committed:
[FIX] Yield session in a clear_upon_failure() environment

npiganeau committed:
[FIX] Set job to failed after non-retryable OperationalError

pedrobaeza committed:
Merge pull request #132 from npiganeau/8.0-fail-nr-operror

[FIX] Set job to failed after non-retryable OperationalError

Commit:
Update mailing list address

Commit:
Update mailing list address

Commit:
Update mailing list address

Olivier-LAURENT committed:
[8.0][FIX] when a job fails, inactive users are added to its followers

Commit:
Context manager to acquire Postgres advisory locks

This Postgres feature is invaluable when dealing with synchronisations,
especially when importing concurrently records from a system.

When we export a record, we are able to acquire a lock on the exported record
to prevent 2 jobs to export it at the same time. This is different when we
import a record for the first time and with several jobs running in parallel,
chances are high that 2 jobs will import the same record at the same moment.

The Postgres advisory lock comes handy there for they allow to acquire an
application lock.  Usually we'll acquire the lock at the beginning of an import
(beginning of ``Importer.run()``) and we'll throw a ``RetryableJobError`` if
the lock cannot be acquired so the job is retried later. The lock will remain
in place until the end of the transaction.

Example:
 - Job 1 imports Partner A
 - Job 2 imports Partner B
 - Partner A has a category X which happens not to exist yet
 - Partner B has a category X which happens not to exist yet
 - Job 1 import category X as a dependency
 - Job 2 import category X as a dependency

Since both jobs are executed concurrently, they both create a record for category X.
With this lock:

 - Job 1 imports Partner A, it puts a lock for this partner
 - Job 2 imports Partner B, it puts a lock for this partner
 - Partner A has a category X which happens not to exist yet
 - Partner B has a category X which happens not to exist yet
 - Job 1 import category X as a dependency, it puts a lock for this category
 - Job 2 import category X as a dependency, try to put a lock but can't, Job 2
   is retried later, and when it is retried, it sees the category X created by
   Job 1

See http://topopps.com/implementing-postgres-advisory-locks/ for the article
where I learned about the computation of the hash for this purpose.

Commit:
Rename try_advisory_lock to advisory_lock_or_retry

It is no longer a context manager, because we would expect the lock to be
released at the end of the 'with' statement but it lasts until the end of the
transaction.

Commit:
Merge pull request #135 from guewen/8.0-ml-url

Update mailing list address

Commit:
Merge pull request #131 from npiganeau/8.0-npiganeau-patch-1

Clear environment caches and recomputations upon failures

oca-transbot committed:
OCA Transbot updated translations from Transifex

maxime-c2c committed:
Update index.rst

maxime-c2c committed:
Update index.rst

maxime-c2c committed:
Update index.rst

mdietrichc2c committed:
Set job back to 'pending' in case of exception

Commit:
Fix bug introduced in c97ebbd in jobrunner

Commit c97ebbd was a frontport from 7.0.
However, openerp.sql_db.dsn was changed in 8.0 in that it now returns a
tuple. Extract the db_name from the returned tuple.

Commit:
Fix backend_to_m2o to extract id of the binding

The default binder returns a recordset of one record since 41561c8 so
backend_to_m2o should extract the Id from it.

Commit:
Fix the fix c3fefd2 :-(

The first item contains the name of the DB such as 'odoo_db' and the
second item contains 'user=gbaconnier dbname=odoo_db' which is what
expects psycopg2.connect.

sbidoul committed:
[FIX] correctly obtain the list of database with odoo is started with --no-database-list

sbidoul committed:
[FIX] ignore dbfilter containing %d or %h, fixes #58

This is a temporary fix. In version 4.0,
dbfilter will be completely ignored by connector.

npiganeau committed:
Manage non-ascii PG errors

When PG is localized, error messages are not ascii and jobs are not postponed. Instead they are failed with a 'unicode decode error'.

Note: even if PG ```lc_messages``` is set to en_us.utf8 on a localized system, the error message prefixes (such as ```DETAILS:```) are still localized.

npiganeau committed:
Removed trailing whitespace

Commit:
Merge pull request #166 from acsone/8.0-fix-dbfilter-sbi

[FIX] ignore dbfilter containing %d or %h, fixes #58

sbidoul committed:
[FIX] ignore dbfilter, fixes #58

sbidoul committed:
[FIX] ignore dbfilter, fixes #58

npiganeau committed:
Use tools.ustr() instead of unicode()

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
Merge pull request #167 from npiganeau/patch-1

Manage non-ascii PG errors

Commit:
Prevent to unpickle globals which are not jobs

This is a safeguard to prevent someone to write arbitrary code in jobs.
Builtin types and datetime/timedelta are allowed in job arguments, and a
new function 'whitelist_unpickle_global' allows to register new objects
if needed.

lmignon committed:
[MIG] Make module installable

lmignon committed:
[MIG] Use python standard lib unittest in place of unittest2

lmignon committed:
[MIG] email followers are instance of model 'mail.followers'

lmignon committed:
[MIG] value returned by _company_default_get is a BrowseRecord

lmignon committed:
[DEL] Remove the old connector worker

lmignon committed:
[PEP8]

lmignon committed:
[FIX] In the last release of Odoo related partner created from an inactive user is also inactive
Fix the test accordingly (https://github.com/odoo/odoo/commit/328d34a22168bd9106dde810981bfd838b675a5c)

lmignon committed:
[FIX] The first version is 9.0.1.0.0

lmignon committed:
[FIX] fix path of odoo for doc build

Commit:
Fix too long line

Commit:
Release 8.0.3.3.0

lmignon committed:
[DEL] Remove deprecated APIs
* In openerp.addons.connector.session.ConnectorSession:
  * Remove method 'search'
  * Remove method 'write'
  * Remove method 'browse'
  * Remove method 'read'
  * Remove method 'create'
  * Remove method 'unlink'
* In openerp.addons.connector.connector.ConnectorUnit:
  * Remove method 'get_connector_unit_for_model'
  * Remove method 'get_binder_for_model'
  * Remove property 'environment'
* In openerp.addons.connector.connector.ConnectorEnvironment:
  * Remove method 'set_lang'
* In openerp.addons.connector.connector.connector.MetaConnectorUnit
  * Remove property 'model_name'
* Remove class openerp.addons.connector.connector.Environment
* Remove method openerp.addons.connector.connector.intall_in_connector
* In openerp.addons.connector.unit.synchronizer:
  * Remove class 'ExportSynchronizer'
  * Remove class 'ImportSynchronizer'
  * Remove class 'DeleteSynchronizer'
* Remove module openerp.addons.connector.deprecate.
*

lmignon committed:
Merge remote-tracking branch 'origin/8.0-no-dbfilter-sbi' into 9.0-from-8.0

YannickB committed:
Add job_uuid in context

sbidoul committed:
Merge pull request #169 from acsone/9.0-from-8.0

9.0 from 8.0

sbidoul committed:
remove remnants of workers

sbidoul committed:
make connector installable

fixes #177

sbidoul committed:
enable the jobrunner by default

fixes #175

sbidoul committed:
bump version

sbidoul committed:
remove unused import

sbidoul committed:
update changelog

sbidoul committed:
adapt to upstream api change to obtain db connection

sbidoul committed:
Bump version number

Commit:
Update changelog for 9.0.1.0.2

oca-transbot committed:
OCA Transbot updated translations from Transifex

nicolas-petit committed:
Added test for job uuid in session context.

nicolas-petit committed:
Removing log traces.

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

lasley committed:
Move connector settings menu to technical

Commit:
Remove cancellation of jobs and their active flag

Now jobs are only set to Done when NothingToDoJob is raised.

Closes #197

yvaucher committed:
[DOC] fix highlight of license

oca-transbot committed:
OCA Transbot updated translations from Transifex

lasley committed:
Merge pull request #201 from nicolas-petit/9.0

Porting job_uuid in context to V9

sbidoul committed:
[FIX] remove en.po that was erroneously created by transbot

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
[FIX] error handling bug

sbidoul committed:
[DOC] remove mention of v9 crowdfunding

sbidoul committed:
[REF] remove old code that was there for backward compatibility

sbidoul committed:
[ADD] first implementation of channel delay

sbidoul committed:
[IMP] to be consistent, use the same 'now' concept everywhere

sbidoul committed:
[IMP] more precise wakeup at the end of delays between jobs

This also runs jobs that have an ETA at a more precise time:
before the runner was waking up only every minute when there was
no other activity.

sbidoul committed:
rename delay key as throttle

Commit:
Merge pull request #205 from acsone/9.0-channel-delay-sbi

[9.0] delay between job start, configurable per channel

Commit:
Fix wrong recordset returned when no binding found

When we call the binder 'to_openerp' method with 'unwrap', we expect the
binder to return the 'unwrapped' record of the binding, e.g.
product.product instead of prestashop.product.product.

When no record is found, the binder *must* return an empty recordset of
the correct (unwrapped) model too.

lasley committed:
Merge pull request #194 from guewen/9.0-fix-backend_to_m2o-binder

Fix backend_to_m2o to extract id of the binding

lasley committed:
[FIX] connector: Zero value assertion
* Switch zero assertion to type strict eval (`is`)

oca-transbot committed:
OCA Transbot updated translations from Transifex

lasley committed:
Merge pull request #198 from guewen/remove-job-active

Remove cancellation of jobs and their active flag

Commit:
Set addons uninstallable

Commit:
Make the connector addon installable

Commit:
Replace 'select=True' by 'index=True'

Commit:
Rename openerp to odoo in Python files

Commit:
Rename xml files with odoo instead of openerp

Commit:
Rename openerp to odoo in docs

Commit:
Replace monkey-patch by a proper inherit

Now that we have a root model 'base', we can properly override the
methods to fire the events.

Commit:
Remove usage of @api.one

Commit:
Replace ConnectorSession by odoo.Environment

Commit:
Upgrade version number

Commit:
Merge JobStorage into Job

The reason of being for the separation was to be able to support another
storage than postgres if needed.  Nothing has ever been done in this
way and the current backend storage (postgres) has proved to be
efficient.

Commit:
Extract the queue_job from connector

Commit:
Remove ConnectorSession

Commit:
Implement DelayableRecordset

The next commits will remove the support of '.delay()' on standalone
methods. It will simplify the code as we now support the 2.

Standalone jobs not on a particular model will need to be done on
AbstractModel created on purpose, so we will only support jobs on
models.

Commit:
Rename misnamed file

Commit:
Fix tests on related actions

Commit:
Encode/Decode recordsets to JSON in job arguments

Commit:
Rework related actions, now methods of Models

The related actions are now methods of the queue.job model.
They are configured using their name instead of being passed as a
function.

Commit:
Fix pylint warnings

Commit:
Use a discriminant field type for JobSerialized

This key is used to register the new field in the fields registry so it
is important to have a unique type.

Commit:
Extract queue_job from OCA/connector

The 'connector' module's part handling the queue_job is split in this
module.

Commit:
Set base repository files

Commit:
pep8 error

Commit:
Add queue_job's readme

Commit:
Use short copyrights

lmignon commented on queue_job/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+          'data/queue_data.xml',
+          ],
+ 'installable': True,
+ 'application': True,
```
Is-it really an application?

guewen replied:
Isn't it an application if we have a top-level menu for accessing the queue models (Channels, Job functions and Jobs). Corollary, would we want a top-level menu? (I'd say yes, and we need to find an icon BTW).

blaggacao replied:
Although they are, I would argue they shouldn't. I feel coining it "application" should account mainly for the business end user perspective. So, If I can't magically hire the best people in the world with this or similar stuff, then it's probably not an "application". This implies abstaining from top-level menu. Maybe it's better classified (hidden) in the "technical" settings menu.

guewen replied:
@blaggacao if the user has not the accesses rights for the queue, he won't see the top-level menu, so for a technical end-user, it can make sense to have a top-level menu (jobs is a menu you open often as you regularly want to check them, it should not be hidden in an obscure place).

lmignon replied:
@guewen your POV makes sense :+1:

leorochael replied:
Even if it does have a top level menu, I'm not sure it's enough to call it an application.

Personally, I'd be ok if the Queue menu was somewhere inside Technical instead of being top-level.

If I need to keep an eye on the jobs queue all the time so that a top-level menu makes sense, then I'd argue something is wrong in my application...

guewen replied:
Then you never tried to synchronize with Magento :joy: .
Joke apart, I'm not sure what exactly is supposed to be an Application. So given the controversy, I removed the flag.

blaggacao replied:
Convincing and true. I was wrong.

lmignon commented:
@guewen What a great refactoring!!!!! No time for a full review yet. After a first read I have not found anything to complain. I need to investigate a little further to completly understand what is behind the changes and for exemple at https://github.com/OCA/queue/pull/1/files#diff-8b63facccbd0e181392cdc06745d034eR615

guewen commented:
I still have to test in _in real_ with `base_import_async` because it's only tested with the python tests for now :)

Commit:
[ADD] Add setup.py

leorochael commented:
As a test, I've tried to migrate a little project that doesn't try to use the whole connector to use just this PR, but quickly ran into an issue: We also use the connector events subsystem...

It seems to me that many uses of the queue without the connector will use (or could benefit from the use of) the events as well.

Should we try to move it along into `queue` or should it perhaps have its own repo?

leorochael commented:
Speaking of moving modules into the queue repo, I think base_import_async would be a good candidate, as it uses solely the job queue.

Rewriting base_import_async to use the new model method delay would be a good way to taste what that api should feel like.

leorochael commented:
And speaking of model methods delay, if we add a method to the global BaseModel object that allows delaying methods, why not simply allow delaying all methods (with serializable parameters) by default?

That's what ERP5 does, by the way.

You could then just add decorators to the methods where you want to pass different queue parameters than the defaults (like priority, or channel name).

guewen commented:
> It seems to me that many uses of the queue without the connector will use (or could benefit from the use of) the events as well.

You might be right. However, it might be done in a subsequent pull request.

guewen commented:
> why not simply allow delaying all methods (with serializable parameters) by default?

I thought about that. Actually, I am _restricting_ to delay methods with the `@job` decorator so it would just need to remove this restriction. 

We should just think about the security. The `with_delay` does a `getattr` on the model, so I used the `@job` decorator as a whitelist of the jobs to run. Before removing the restriction, we must be sure that we don't open a security hole.

The other point is that currently we show the "job functions" in a view so we can configure the channels. It might be weird to have only some of them, but maybe that's just a matter of improving the way we present and name them.

lmignon commented:
> It seems to me that many uses of the queue without the connector will use (or could benefit from the use of) the events as well.

IMO the events subsystem is also a good candidate for a generic module not related at all to the connector module nor the queue_job module.

leorochael commented:
@lmignon I agree, but even as an independent module, should it be in a separate repo?

lmignon commented:
@leorochael IMO yes it should be in a separate repo with its own release management/cycle.

Commit:
Remove the application flag

Commit:
[FIX] Import jobrunner

Commit:
[FIX] Use env

Commit:
[FIX] Add the possibility to store a job with a new cursor.

Commit:
[FIX] im_class seems not accessible

Commit:
[FIX] __inner__ seems not accessible

Commit:
[FIX] At this step, the class isn't instanciated

adrienpeiffer commented:
Hi @guewen,

Thanks for the work !
I did some functional test on this one. I just created a PR (https://github.com/guewen/queue/pull/2) on your own branch to fix some things that appeared during my tests.

We can discuss on each of my commits if you want.

Thanks,

Adrien

Commit:
Merge pull request #1 from acsone/10-queue-job-setup-ape

[ADD] Add setup.py

Commit:
Revert "[FIX] Add the possibility to store a job with a new cursor."

This reverts commit 70c2713769cbabcdc9ea75d07818434b2dfd8cfb.

Commit:
[IMP] Improve the way to store job after exception

Commit:
Merge pull request #2 from acsone/10.0-queue-job-ape

Some proposal after functional testing

Commit:
fix pep8

Commit:
[FIX] Give the db name to the request's session

Commit:
Make tests pass

Separate tests that need special methods in a dedicated test_queue_job
module.

Rework the registry of the jobs. The 'job' decorator is applied on
functions before they are bound to their class, so we don't know the
class yet. Now, the decorator marks job methods with 'delayable' and
when a model is registered (Base._register_hook), it looks for the
marked methods and register them.

Commit:
Fix implementation of related actions

guewen commented:
@adrienpeiffer @lmignon @sbidoul @leorochael @blaggacao  The last 2 commits rework the final bits of the jobs and related actions implementation. Tests are green on my computer. Waiting for the Travis build. Hopefully merged in 2016? :-)

guewen commented:
Aaaand it's green!

parthivgls commented:
👍 +1

sbidoul commented on queue_job/controllers/main.py:
```diff
@@ -0,0 +1,132 @@
+        http.request.env.cr.commit()
+
+        _logger.debug('%s started', job)
+        job.perform()
```
Don't we need some kind of sudo(job.user_id) here?

guewen replied:
the sudo is done in `perform()` that executes `self.func()`, the latter being the method call on the sudoed recordset: https://github.com/guewen/queue/blob/43bb875adc421f6773a07ae6ba143014d7a0e600/queue_job/job.py#L419

sbidoul commented on queue_job/controllers/main.py:
```diff
@@ -0,0 +1,132 @@
+            job.env.clear()
+            with odoo.api.Environment.manage():
+                with odoo.registry(job.env.cr.dbname).cursor() as new_cr:
+                    new_env = api.Environment(new_cr, job.env.uid,
```
job.env.uid is always SUPERUSER_ID

guewen replied:
Do you mean we should replace `job.env.uid` by `SUPERUSER_ID`? I replaced by `job.env(cr=new_cr)` which returns the new env with only the `cr` modified, avoiding some repetition.

sbidoul commented on queue_job/__manifest__.py:
```diff
@@ -0,0 +1,22 @@
+
+{'name': 'Job Queue',
+ 'version': '10.0.1.0.0',
+ 'author': 'Camptocamp,Acsone,Odoo Community Association (OCA)',
```
s,Acsone,ACSONE SA/NV,

guewen replied:
Fixed

sbidoul commented on queue_job/models/base.py:
```diff
@@ -0,0 +1,58 @@
+                       in inspect.getmembers(self, predicate=inspect.ismethod)
+                       if getattr(method, 'delayable', None)]
+        for job_method in job_methods:
+            self.env['queue.job.function']._register_job(job_method)
```
@guewen nice!

Commit:
Correct ACSONE SA/NV wording

Commit:
Simplify usage of Environment with new cr

Commit:
Move tests using queue_job.testing_method in test_queue_job

To avoid to pollute 'queue.job.function' with a test job.

sbidoul reviewed (approved):
Great job!

leorochael commented:
Hi @guewen, one thing I'm missing in this PR is the configuration file effort I did for `connector` on OCA/connector#213.

Since there wasn't a maintained 10.0 branch of connector yet, I ended up not porting it over before your queue_job extraction effort, even though it was already landed on 9.0 and it would obviously be trivial to make it work on 10.0.

That PR, along with odoo/odoo#13685, which got squash merged on odoo/odoo@5ee4344, means it should be possible to configure and launch a `queue_job` enabled odoo 10.0 entirely from `odoo.cfg` without having to resort to environment variables.

I'd like to see the config file work land on queue_job before this PR is merged, but I'm not sure when I'm going to find the [round tuits](https://en.wiktionary.org/wiki/round_tuit) to make it happen on my side, and I don't want to hold this PR hostage, so feel free to land it without.

The worse that could happen is that I'll make another PR later...

adrienpeiffer commented:
Hi Guewen,
Can you take a look on https://github.com/guewen/queue/pull/3 ?

Commit:
Merge pull request #3 from acsone/10.0-queue-job-ape

Give the db name to the request's session

adrienpeiffer commented:
Hi,

One last thing, I noticed that when an object in memory that is in parameter of the delayed method, and that this object is modified (like a dictionary). When the job fails, changes made to the dictionary remain persistent. Any comment ?

pedrobaeza commented on queue_job/README.rst:
```diff
@@ -0,0 +1,167 @@
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/{repo_id}/{branch}
```
Put this information

pedrobaeza commented on queue_job/README.rst:
```diff
@@ -0,0 +1,167 @@
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/{repo_id}/{branch}
+
+.. repo_id is available in https://github.com/OCA/maintainer-tools/blob/master/tools/repos_with_ids.txt
```
Remove this

guewen commented:
@adrienpeiffer could you give an example?

leorochael commented:
@adrienpeiffer, this is normal behaviour in Python,  if you pass mutable objects into a function/method and the function mutates the object, the object remains mutated outside of the scope of the function, even if the function raises an exception.

Only ORM records know to revert state when a transaction is aborted.

The solution is either not to mutate the object in the delayed call, or to make a copy if you need to.

leorochael commented:
@guewen: guewen/queue#5 for the config file work.

adrienpeiffer commented:
@leorochael Of course. But in this case, these mutable objects are stored in the database when the job failed. So, I think that it's not really clean.

@guewen. For instance, I give a dictionary that contains all data necessary the create a new partner. When the job failed (for any reason) the dictionary (args, kwargs) of the delayed method contains other key 
in the dictionary like ```message_followers_ids``` or something like that.

Personally, I just make a deep copy of the dictionary in the delayed method when the job is running.

Commit:
Fix mutability of job arguments

The model name, method name, recordset ids, arguments and keyword
arguments must never ever be modified after the creation of a job.
This change fix a bug when a job with a mutable argument see the content
of the argument being modified during the execution of a job.

guewen commented:
@adrienpeiffer Nice catch! Thanks for the explanation, I could fix it in efddad9.

@leorochael Thanks for bringing this feature back. I added a comment on the pull request regarding the naming of the section `options-queue_job`.

Commit:
Complete OCA runbot url

sbidoul committed:
[IMP] sequential channels from 9.0

sbidoul commented:
And when https://github.com/guewen/queue/pull/5 is merged, I'll do a PR for the sequential channels feature.

leorochael committed:
Allow configuring runner from config file

In addition to the ODOO_CONNECTOR_CHANNELS environment variable, for
backward compatibility, we can configure the runner channels through
the ODOO_QUEUE_JOB_CHANNELS and in the odoo configuration file like this:

    [options-job_queue]
    channels = root:4

For backward compatibility, the section `[options-connector]` is also
accepted.

Moreover, it's now also possible to preload the module in the config
file, like:

    [options]
    server_wide_modules = web,web_kanban,queue_job

As such we can configure the job runner using only the Odoo config file.

leorochael committed:
Whitespace handling on channel configuration

Ignore whitespace around values, and tolerate missing entries that would
be caused by trailing commas or commented lines when the channel
configuration is provided through the Odoo configuration file.

Also, accept line breaks channel entry separators along with commas,
which make the configuration file more readable.

adrienpeiffer reviewed (approved):
:+1: Great work !

leorochael committed:
Remove all references to CONNECTOR

Or replace them with references to queue_job.

Also, rip off the barely useful modicum of backward compatibility.

Commit:
Merge pull request #5 from leorochael/10.0-add-queue_job-config-file

Forward port config file setup from connector

sbidoul committed:
[FIX] pep8 plus trivial doctest fix

sbidoul commented:
@guewen here is the last piece of the puzzle: https://github.com/guewen/queue/pull/6

Commit:
Merge pull request #6 from acsone/10.0-seq-channels-sbi

10.0 seq channels sbi

guewen commented:
Thanks @sbidoul. I'm eager to see this merged :)

sbidoul commented:
Voilà @guewen. À toi l'honneur du tweet.

blaggacao commented:
@guewen Congratulations! This is actually really exciting work. To give people like me a headstart, consider a possible use cases section in the readme? Like advertising it a little bit, in the words of a seasoned contributor?

leorochael commented:
We should consider adding a github-pages or ReadTheDocs auto-generated site next.

Also, we should probably port [base_import_async](https://github.com/OCA/connector-interfaces/tree/8.0/base_import_async) (and perhaps move it into this repo) as an example of what can be done with `queue_job`.
